### PR TITLE
Create custom document preview generator doc

### PR DESCRIPTION
**Pull request by tacoelho**

### CODE REVIEW

File: docs/custom-document-preview-generator.md [hunk 1]
```diff
@@ -0,0 +1,210 @@
+Overview
```
Comment (herflis) [inline docs/custom-document-preview-generator.md]:
```suggestion
---
title: "Create a custom document preview generator"
source_url: 'https://github.com/SenseNet/sn-preview/blob/master/docs/custom-document-preview-generator.md'
category: Development
version: v7.0
tags: [document, preview, preview generator]
---
Overview
```

File: docs/custom-document-preview-generator.md [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+========
+
+**Sense/Net ECM** provides means to [view documents in the browser](http://wiki.sensenet.com/Viewing_documents_in_the_browser "Viewing documents in the browser") without the need of *Microsoft Office* or other tools to be installed on the users' computers. The built-in solution is only available in the **Enterprise Edition**. However Sense/Net developers can integrate any third-party preview generator module easily even in the **Community Edition**. This tutorial shows you how to do it.
```
Comment (herflis) [inline docs/custom-document-preview-generator.md]:
```suggestion
**sensenet** provides means to [view documents in the browser](http://wiki.sensenet.com/Viewing_documents_in_the_browser "Viewing documents in the browser") without the need of *Microsoft Office* or other tools to be installed on the users' computers. The built-in solution is only available in the **Enterprise Edition**. However sensenet developers can integrate any third-party preview generator module easily even in the **Community Edition**. This tutorial shows you how to do it.
```

File: docs/custom-document-preview-generator.md [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+
+The information in this article is related to **Sense/Net ECM 6.3.1** and above. For implementations for previous versions please refer to the source code.
+
+Creating the provider
+=====================
+
+To implement a custom preview provider you will need to [create a new class](http://wiki.sensenet.com/How_to_create_a_new_class_for_source_code "How to create a new class for source code") that inherits one of the following base classes:
```
Comment (herflis) [inline docs/custom-document-preview-generator.md]:
```suggestion
To implement a custom preview provider you will need to [create a new class](https://github.com/SenseNet/sensenet/blob/master/docs/how-to-create-a-new-class-for-source-code.md) "How to create a new class for source code") that inherits one of the following base classes:
```

File: docs/custom-document-preview-generator.md [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+
+Generating images
+=================
+
+The preview provider uses the [Task Management](http://wiki.sensenet.com/Task_Management "Task Management") framework for queueing and managing image generator tasks. The actual preview image generation is done using a command line tool (a task tool), as you can see later in this article
+
+In your provider you have to override at least the following methods and properties:
+
+-   **IsContentSupported(Node)**: returns if this implementation supports the specified content. This is necessary for performance reasons: the task that will generate images will only start for supported content (e.g. documents, but not Javascript files).
+-   **GetPreviewGeneratorTaskName(path)**: this method should return the name of the preview generator task (the name of the command line tool). Please read the [Task Management](http://wiki.sensenet.com/Task_Management "Task Management") article for more details about tasks.
```
Comment (herflis) [inline docs/custom-document-preview-generator.md]:
```suggestion
-   **GetPreviewGeneratorTaskName(path)**: this method should return the name of the preview generator task (the name of the command line tool). Please read the [Task Management](https://github.com/SenseNet/sn-taskmanagement/blob/master/docs/task-managemet.md "Task Management") article for more details about tasks.
```

File: docs/custom-document-preview-generator.md [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+
+Before version **6.3.1 Patch 4** this is a property; after the patch it is a method that may return a different task name based on the content. This may be useful when you plan to execute certain tasks on dedicated agents.
+
+If you only want to integrate a custom preview generator tool, but leave the built-in preview serving mechanism as-is, then your custom provider is done - please continue with the [image generator tool](http://wiki.sensenet.com/How_to_integrate_a_custom_document_preview_generator#Creating_the_image_generator_tool) section below.
```
Comment (herflis) [inline docs/custom-document-preview-generator.md]:
```suggestion
If you only want to integrate a custom preview generator tool, but leave the built-in preview serving mechanism as-is, then your custom provider is done - please continue with the image generator tool section below.
```

File: docs/custom-document-preview-generator.md [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+		...
+    }
+}
+
+The class can be anywhere in your project as it will be discovered automatically by the system when the application starts.
+
+Generating images
+=================
+
+The preview provider uses the [Task Management](http://wiki.sensenet.com/Task_Management "Task Management") framework for queueing and managing image generator tasks. The actual preview image generation is done using a command line tool (a task tool), as you can see later in this article
```
Comment (herflis) [inline docs/custom-document-preview-generator.md]:
```suggestion
The preview provider uses the [Task Management](https://github.com/SenseNet/sn-taskmanagement/task-management.md "Task Management") framework for queueing and managing image generator tasks. The actual preview image generation is done using a command line tool (a task tool), as you can see later in this article
```

File: docs/custom-document-preview-generator.md [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+-   **EmptyPreviewsFolder(Node)**: clears all preview images in a preview folder. Override this if you want to store images in a different place than the built-in one.
+-   **RemovePreviewImagesAsync(int, VersionNumber, bool)** (from version **6.3.1 Patch 3**): Deletes preview images for one version or for all of them asynchronously.
+
+For additional helper methods available in the base class (e.g. getting preview or thumbnal image names or modifying the page count or status of the document) please refer to the source code.
+
+Restricted preview images
+=========================
+
+It is possible to add restrictions to the displayed preview images. See the [document preview article](http://wiki.sensenet.com/Viewing_documents_in_the_browser#Restrictions_and_annotations "Viewing documents in the browser") for detailed explanation. The built-in document preview provider has a default implementation for serving restricted images, but you can customize that behavior as shown in the following sections.
```
Comment (herflis) [inline docs/custom-document-preview-generator.md]:
```suggestion
It is possible to add restrictions to the displayed preview images. See the [document preview article](/viewing-documents-in-the-browser.md "Viewing documents in the browser") for detailed explanation. The built-in document preview provider has a default implementation for serving restricted images, but you can customize that behavior as shown in the following sections.
```

File: docs/custom-document-preview-generator.md [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+You can override the **GetRestrictionType** method of DocumentPreviewProvider to implement custom logic for determinig whether the user has the appropriate permissions for preview images. The method returns a **flag enumeration value** that contains the restrictions that need to be placed onto the preview image. The portal uses this method to determine if the user has any previw permissions. If not, the portal will not serve preview images.
+
+The possible return values are the following:
+
+-   **NoAccess**: the user does not have any preview permissons. If this value is returned, the portal will not serve preview images to the user.
+-   **NoRestriction**: the user has access to the full preview, without restrictions.
+-   **Redaction**: redactions must be placed onto the document if they exist.
+-   **Watermark**: a watermark must be placed onto the document if it exists.
+
+The built-in document preview provider works with the built-in [preview permissions](http://wiki.sensenet.com/Viewing_documents_in_the_browser#Preview_permissions "Viewing documents in the browser"). You can customize this method for example to check group membership instead:
```
Comment (herflis) [inline docs/custom-document-preview-generator.md]:
```suggestion
The built-in document preview provider works with the built-in [preview permissions](/viewing-documents-in-the-browser.md "Viewing documents in the browser"). You can customize this method for example to check group membership instead:
```

File: docs/custom-document-preview-generator.md [hunk 9]
```diff
@@ -0,0 +1,210 @@
...
+        }
+    }
+}
+
+Creating the image generator tool
+=================================
+
+If you want to use the built-in preview generator shipped with the **Enterprise Edition**, you do not have to create your own generator. This is necessary only in case of the **Community Edition** or if you want to replace the built-in one.
+
+Preview image generation is done using a [task executor tool](http://wiki.sensenet.com/Task_Management#Creating_a_custom_task_executor "Task Management"). This is a command line tool that downloads the document (e.g. a word document), and generates preview images using a 3rd party image generator plugin (e.g. [Aspose](http://www.aspose.com/)).
```
Comment (herflis) [inline docs/custom-document-preview-generator.md]:
```suggestion
Preview image generation is done using a [task executor tool](https://github.com/SenseNet/sn-taskmanagement/task-management.md "Task Management"). This is a command line tool that downloads the document (e.g. a word document), and generates preview images using a 3rd party image generator plugin (e.g. [Aspose](http://www.aspose.com/)).
```

File: docs/custom-document-preview-generator.md [hunk 10]
```diff
@@ -0,0 +1,210 @@
...
+
+### IPreviewGenerationContext interface
+
+Provides properties and methods for context information in a specific generator implementation. This context is responsible for holding document-specific information and for the communication with the portal (e.g. setting the page count or preview status of the document and saving the generated images to the portal). In most cases you will not need to change the built-in implementation, only if you want to customize the environment of the preview generation process. The methods and properties defined in this interface are the following.
+
+-   **ContentId**: id of the document
+-   **PreviewsFolderId**: id of the preview folder where images will be saved
+-   **StartIndex**: the page number where we need to start generating images
+-   **MaxPreviewCount**: maximum number of images to generate
+-   **PreviewResolution**: [image resolution](http://wiki.sensenet.com/Viewing_documents_in_the_browser#Generator_configuration "Viewing documents in the browser")
```
Comment (herflis) [inline docs/custom-document-preview-generator.md]:
```suggestion
-   **PreviewResolution**: [image resolution](/viewing-documents-in-the-browser.md "Viewing documents in the browser")
```

File: docs/custom-document-preview-generator.md [hunk 11]
```diff
@@ -0,0 +1,210 @@
...
+-   **SaveEmptyPreview**: saves an empty preview image in case the 3rd party image generator was not able to recognise a page.
+-   **SaveImage**: a different API for saving both preview and thumbnail images using the previously defined *SavePreviewAndThumbnail* method.
+-   **LogInfo**: logs an information message
+-   **LogWarning**: logs a warning message
+-   **LogError**: logs an error message
+
+OData API for the image generator
+---------------------------------
+
+The tool uses the [OData REST API](http://wiki.sensenet.com/OData_REST_API "OData REST API") of Sense/Net to get the necessary information from the portal and upload the generated preview images. The Document Preview Provider has the following **OData API** (OData actions and functions) for this tool:
```
Comment (herflis) [inline docs/custom-document-preview-generator.md]:
```suggestion
The tool uses the [OData REST API](https://github.com/SenseNet/sensenet/blob/master/docs/odata-rest-api.md "OData REST API") of sensenet to get the necessary information from the portal and upload the generated preview images. The Document Preview Provider has the following **OData API** (OData actions and functions) for this tool:
```

File: docs/custom-document-preview-generator.md [hunk 12]
```diff
@@ -0,0 +1,210 @@
...
+
+-   **GetPreviewsFolder**: loads or creates the preview folder for the given version of the content. Optionally it can also empty the folder in case a cleanup is needed.
+-   **SetPreviewStatus**: sets the preview status of the document (e.g. Error, NotSupported or InProgress)
+-   **SetPageCount**: saves the page count of a document as a metadata field.
+-   **SetInitialPreviewProperties**: a helper method that sets the initial preview image properties necessary for the preview framework to work (e.g. the creator or modifyer user). It is mandatory to call this method after creating a new preview image.
+
+To actually upload the images you may use the built-in [Upload action](http://wiki.sensenet.com/Upload_action "Upload action"). See the example in that article about how to upload image binaries to the portal.
```
Comment (herflis) [inline docs/custom-document-preview-generator.md]:
```suggestion
To actually upload the images you may use the built-in Upload action.
```

File: docs/custom-document-preview-generator.md [hunk 13]
```diff
@@ -0,0 +1,210 @@
...
+
+Deploying the tool
+------------------
+
+As the tool is a task tool, please read the [following article](http://wiki.sensenet.com/Task_Management#Deploying_the_task_executor "Task Management") about deploying it.
```
Comment (herflis) [inline docs/custom-document-preview-generator.md]:
```suggestion
As the tool is a task tool, please read the [following article](https://github.com/SenseNet/sn-taskmanagement/blob/master/docs/task-managemet.md "Task Management") about deploying it.
```

**Comment by herflis** [comment]
Hi @tacoelho ,
I made some minor suggetions, please commit them, but otherwise it is an awesome work, thank you for your contribution! 🙏

**Comment by tacoelho** [comment]
> Hi @tacoelho ,
> I made some minor suggetions, please commit them, but otherwise it is an awesome work, thank you for your contribution!

My pleasure!

**Pull request merged**